### PR TITLE
Update basic02

### PR DESCRIPTION
**Pull request by matsduf**
Clarify different types of errors that the test case can detect so that better messages can be sent to the users. More explicit steps. The updated version is compatible with the previous version. No know changes in steps compared to implementation.

### CODE REVIEW

File: docs/specifications/tests/Basic-TP/basic02.md [hunk 1]
```diff
@@ -1,45 +1,65 @@
...
-### Special procedural requirements
+## Inputs
+
+The label of the domain name to be tested (the child zone).
+
+## Ordered description of steps to be taken to execute the test case
+
+1. Retrieve the NS records for the child zone (the delegation) using
+   [Method 2]. If the test is an undelegated test, use the name servers
+   from input data.
```
Comment (mattias-p) [inline docs/specifications/tests/Basic-TP/basic02.md]: Here you're introducing a reference to input data. Please add an item for this in the Inputs section. Give it a name and refer to that here.

File: docs/specifications/tests/Basic-TP/basic02.md [hunk 2]
```diff
@@ -1,45 +1,76 @@
...
+## Objective
 
 In order for the domain to work, it must have at least one name server that
 can answer queries about the domain. 
 
-### Inputs
+## Inputs
+
+* The label of the domain name to be tested ("child zone").
+* If undelegated test, the list of name servers ("undelegated NS").
```
Comment (mattias-p) [inline docs/specifications/tests/Basic-TP/basic02.md]: I suggest clarifying that this list represents the delegation for the child zone.

File: docs/specifications/tests/Basic-TP/basic02.md [hunk 3]
```diff
@@ -1,45 +1,76 @@
...
+2. If the NS set is empty (no delegation), this test case fails
+   ("domain name does not exist").
+3. Retrieve the IP addresses (glue records) for any [in-bailiwick] name
+   servers using [Method 4]. If the test is an undelegated test, use 
+   _undelegated glue IP_.
+4. Retrieve the IP addresses for any [out-of-bailiwick] name servers
+   using recursive queries. If the test is an undelegated test, use 
+   _undelegated non-glue IP_ all such name servers that have IP 
+   address data and do a recursive lookup for the rest.
+5. If no IP addresses could be trieved, this test case fails ("no
```
Comment (mattias-p) [inline docs/specifications/tests/Basic-TP/basic02.md]: "trieved" -> "retrieved"

File: docs/specifications/tests/Basic-TP/basic02.md [hunk 4]
```diff
@@ -1,45 +1,76 @@
...
+  NS, if any ("undelegated glue IP").
+* If undelegated test, the IP addresses of the [out-of-bailiwick]
+  undelegated NS, if any ("undelegated non-glue IP").
+
+## Ordered description of steps to be taken to execute the test case
+
+1. Retrieve the NS records for _the child zone_ (the delegation) using
+   [Method 2]. If the test is an undelegated test, use _undelegated
+   NS_.
+2. If the NS set is empty (no delegation), this test case fails
```
Comment (mattias-p) [inline docs/specifications/tests/Basic-TP/basic02.md]: Maybe use the same language here as in step 6? I.e. "If no NS records could be retrieved".

File: docs/specifications/tests/Basic-TP/basic02.md [hunk 5]
```diff
@@ -1,45 +1,76 @@
...
+   servers using [Method 4]. If the test is an undelegated test, use 
+   _undelegated glue IP_.
+4. Retrieve the IP addresses for any [out-of-bailiwick] name servers
+   using recursive queries. If the test is an undelegated test, use 
+   _undelegated non-glue IP_ all such name servers that have IP 
+   address data and do a recursive lookup for the rest.
+5. If no IP addresses could be trieved, this test case fails ("no
+   IP addresses to name servers").
+6. Create an NS query for the apex of _the child zone_.
+7. Send the query to each available name server IP address.
```
Comment (mattias-p) [inline docs/specifications/tests/Basic-TP/basic02.md]: "Send the query" -> "Send the NS query"?

**Comment by matsduf** [comment]
@sandoche2k, have you reviewed this PR?

File: docs/specifications/tests/Basic-TP/basic02.md [hunk 6]
```diff
@@ -1,45 +1,77 @@
...
-### Objective
+## Objective
 
 In order for the domain to work, it must have at least one name server that
 can answer queries about the domain. 
 
-### Inputs
+## Inputs
+
+* The label of the domain name to be tested ("child zone").
```
Comment (mattias-p) [inline docs/specifications/tests/Basic-TP/basic02.md]: "The label of the domain name" -> "The domain name"?

File: docs/specifications/tests/Basic-TP/basic02.md [hunk 7]
```diff
@@ -1,45 +1,77 @@
...
+   address data and do a recursive lookup for the rest.
+5. If no IP addresses could be retrieved, this test case fails ("no
+   IP addresses to name servers").
+6. Create an NS query for the apex of _the child zone_.
+7. Send the NS query to each available name server IP address.
+8. If there is no response on any query this test case fails 
+   ("no response from name servers").
+9. If there is no valid response containing the NS records of 
+   _the child zone_ this test case fails 
+   ("error in response from name servers").
```
Comment (mattias-p) [inline docs/specifications/tests/Basic-TP/basic02.md]: How is the validity check to be performed?
If there are NS records in the response, how do you know if they are "the NS records of the child zone"?

File: docs/specifications/tests/Basic-TP/basic02.md [hunk 8]
```diff
@@ -1,45 +1,61 @@
...
+1. Retrieve the NS records for the child zone (the delegation) using
+   [Method 2]. If the test is an undelegated test, use the name servers
+   from input data. 
+2. If the NS set is empty (no delegation), this test case fails
+   ("domain name does not exist").
+3. Retrieve the IP addresses (glue records) for all in-bailiwick name
+   servers using [Method 4]. For out-of-bailiwick name servers, do 
+   recursive queries to retrieve the IP addresses.
+4. If no IP addresses could be trieved, this test case fails ("no
+   IP addresses to name servers").
```
[the remark this answers is not in the document]
Comment (matsduf) [reply docs/specifications/tests/Basic-TP/basic02.md]: It is not possible to match directly from RCODE to result. The messages are there to make it more explicit of the various error cases, which was the original purpose of my issue dotse/zonemaster-engine#259.

File: docs/specifications/tests/Basic-TP/basic02.md [hunk 9]
```diff
@@ -38,9 +38,13 @@ can answer queries about the domain.
...
 8. If there is no response on any query this test case fails 
    ("no response from name servers").
-9. If there is no valid response containing the NS records of 
-   _the child zone_ this test case fails 
-   ("error in response from name servers").
+9. At least one name server IP must meet the following requirements or
+   else this test case fails ("error in response from name servers"):
+   1. The response must be a valid DNS response.
+   2. The response must have at least one NS record.
+   3. The owner name of the NS record or records must be identical
```
[the remark this answers is not in the document]
Comment (matsduf) [reply docs/specifications/tests/Basic-TP/basic02.md]: All NS must have the owner name of the child zone. And I wanted to write that we should only look at the NS in the answer section, but I am not sure that is tested today.

**Pull request merged**